File: sunflower-core/src/main/java/org/tdf/sunflower/consensus/pow/Genesis.java
package org.tdf.sunflower.consensus.pow;

import com.fasterxml.jackson.databind.JsonNode;
import org.tdf.common.util.HexBytes;
import org.tdf.sunflower.types.AbstractGenesis;
import org.tdf.sunflower.types.Block;
import org.tdf.sunflower.types.Header;

public class Genesis extends AbstractGenesis {
    public Genesis(JsonNode parsed) {
        super(parsed);
    }

    public HexBytes getNbits() {
        JsonNode n = parsed.get("nbits");
        return n == null ?
            HexBytes.fromHex("ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff") :
            HexBytes.fromHex(n.asText());
    }

    public Block getBlock() {
        if (getNbits().size() != 32)
            throw new RuntimeException("invalid nbits size should be 32");

        Header h = Header.builder()
            .hashPrev(getParentHash())
            .createdAt(getTimestamp())
            .build();

        return new Block(h);
    }
}

File: sunflower-core/src/main/java/org/tdf/sunflower/consensus/pow/Genesis.kt
package org.tdf.sunflower.consensus.pow

import com.fasterxml.jackson.databind.JsonNode
import org.tdf.common.util.HexBytes
import org.tdf.common.util.hex
import org.tdf.sunflower.types.AbstractGenesis
import org.tdf.sunflower.types.Block
import org.tdf.sunflower.types.HeaderImpl

class Genesis(parsed: JsonNode) : AbstractGenesis(parsed) {
    companion object {
        val MAX_N_BITS: HexBytes = "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff".hex()
    }

    val nbits: HexBytes
        get() = parsed["nbits"]?.asText()?.hex() ?: MAX_N_BITS

    override val block: Block
        get() {
            if (nbits.size != 32) throw RuntimeException("invalid nbits size should be 32")
            val h = HeaderImpl(
                hashPrev = parentHash,
                createdAt = timestamp
            )
            return Block(h)
        }
}